test/serverExplorer.test.ts

```diff
@@ -0,0 +1,288 @@
+
+    test('serverExplorer.addServerOutput call should show ServerOutput', () => {
+        const getStub = sandbox.stub(serverExplorer.serverOutputChannels, 'get').returns({
+            show: () => {},
+            append: () => {}
+        });
+        serverExplorer.addServerOutput(ProcessOutput);
+        expect(getStub).calledOnce;
+    });
+
+    test('serverExplorer.refresh() should trigger serverExporer.getChildren() call for root node', () => {
+        const fireStub = sandbox.stub(EventEmitter.prototype, 'fire');
+        serverExplorer.refresh(serverHandle);
+        expect(fireStub).calledOnce;
+        expect(fireStub).calledOnceWith(serverHandle);
+    });
+});
+
+suite('serverExplorer.updateServer', () => {
+
+    let sandbox: sinon.SinonSandbox;
+    let serviceStub: sinon.SinonSandbox;
+    let serverStatusStub: sinon.SinonSandbox;
+    let getStub: sinon.SinonSandbox;
+    const clientStub: sinon.SinonStubbedInstance<RSPClient> = new RSPClient('somehost', 8080);
+    let serverExplorer: ServersViewTreeDataProvider;
+
+    const serverType: Protocol.ServerType = {
+        description: 'a type',
+        id: 'type',
+        visibleName: `the type`
+    };
+
+    const serverHandle: Protocol.ServerHandle = {
+        id: 'id',
+        type: serverType
+    };
+
+    const serverStop = {
+        collapsibleState: 0,
+        label: `id:the type(Stopped)`,
+        contextValue: 'Stopped'
+    };
+
+    const serverStart = {
+        collapsibleState: 0,
+        label: 'id:the type(Started)',
+        contextValue: 'Started'
+    };
+
+    const serverUnknown = {
+        collapsibleState: 0,
+        label: 'id:the type(Unknown)',
+        contextValue: 'Unknown'
+    };
+
+    setup(() => {
+        serverExplorer = new ServersViewTreeDataProvider(clientStub);
+        sandbox = sinon.createSandbox();
+        getStub = sandbox.stub(serverExplorer.serverOutputChannels, 'get').returns({
+            clear: () => {}
+        });
+        sandbox.stub(serverExplorer.servers, 'values').returns(Array.from([serverHandle]));
+    });
+
+    teardown(() => {
+        sandbox.restore();
+    });
+
+    test('call should update server state to received in state change event (Stopped)', () => {
+        const stateChangeStopping: Protocol.ServerStateChange = {
+            server: serverHandle,
+            state: 3
+        };
+        serviceStub = sandbox.stub(serverExplorer.servers, 'get').returns({
+            stateChangeStopping
+        });
+        serverStatusStub = sandbox.stub(serverExplorer.serverStatus, 'set');
+        sandbox.stub(serverExplorer.serverStatusEnum, 'get').returns('Stopped');
+        const stateChangeStopped: Protocol.ServerStateChange = {
+            server: serverHandle,
+            state: 4
+        };
+        const getChildren = serverExplorer.getChildren(undefined);
+        const getTreeItem = serverExplorer.getTreeItem(serverHandle);
+        serverExplorer.updateServer(stateChangeStopping);
```

Comment:
you should not rely on number of calls here, because you are not testing API or framework where it is important. You should merely test that state was actually changed. You should stub everything you need before test. Then you add a server instance and check it is in stopped state. Then emit state changed to starting and check that server changed its state to starting and etc through out all possible states stopped->starting -> started -> stopping -> stopped.